Consolidation of ELF generating capabilities

nbhuiyan opened the pull request:
### Background
When support was added to generate relocatable ELF files, it resulted in a lot of code duplication with the existing executable ELF generating support provided by OMR::CodeCacheManager.

### Changes

This PR introduces 3 new classes:
* ELFGenerator - an abstract base class providing a way to share common code
* ELFExecutableGenerator - Derived from ELFGenerator and handles the generating of executable ELF files, which was previously done entirely by CodeCacheManager
* ELFRelocatableGenerator - Derived from ELFGenerator and handles the generating of relocatable ELF files, which was previously done by the ELFObjectFileGenerator class

Some major changes in OMR::CodeCacheManager include:
* Removal of the functions that laid out executable ELF file layout in memory, and the function that wrote to file.
* Registering of compiled methods and static relocations are handled entirely by CodeCacheManager. Previously, ELFObjectFileGenerator had to deal with what is now moved to CodeCacheManager.
* New structures added to track symbols and relocations to write to ELF files, and keep track of the number of symbols, total length of the symbol names, etc.

Relocatable and executable ELF files were often being referred
to by terms such as "ELF Object File" and "ELF Object" respectively.
This can potentially cause confusions. These changes address that
problem and make a clearer distinction between the 2 variants. Part of
the changes include OMROptions now using
"enableRelocatableELFGeneration" instead of
"enableObjectFileGeneration", as well as a new option
to explicitly enable executable ELF file generation without having
to enable the option TR_PerfTool.

Also adds a documentation about the ELF generating support in OMR Compiler.

Unrelated to what this PR does, there is also a Markdown comment formatting fix that
I noticed when adding the ELFGenerator documentation in the same directory as that document.

### Do not merge until all the boxes have been filled:
- [x] Implement ELFGenerator, ELFExecutableGenerator and ELFRelocatableGenerator
- [x] Update Makefiles for JitBuilder and Test Compiler
- [x] Update CMakeList
- [x] Update OMR::CodeCacheManager to use the new generator classes
- [x] Move away from using structs to set the ELF file layout in memory, and use independent members (ELF64_Ehdr, ELF64_Shdr, etc.) that can be pieced together when writing to file, allowing greater flexibility and code sharing
- [x] Markdown documentation about ELF generating support in Compiler Technology in `doc/compiler`
- [x] Clean removal of ELFObjectFileGenerator class, and updating makefiles accordingly
- [x] Open PR in OpenJ9 with the necessary Makefile update (eclipse/openj9#1689)

nbhuiyan commented:
@0xdaryl The changes so far can be reviewed now.

nbhuiyan commented:
@xliang6

0xdaryl commented on compiler/runtime/OMRCodeCacheManager.hpp:
```diff
@@ -264,21 +257,33 @@ class OMR_EXTENSIBLE CodeCacheManager
+    * @param
+    * @return
+   */
+   void initializeELFGenerator(void);
```
For clarity, I think this should be `initializeExecutableELFGenerator`

0xdaryl commented on compiler/runtime/OMRCodeCacheManager.cpp:
```diff
@@ -970,10 +980,10 @@ OMR::CodeCacheManager::repositoryCodeCacheCreated()
 #if (HOST_OS == OMR_LINUX)
    TR::CodeCacheConfig &config = self()->codeCacheConfig();
    if (config.emitElfObject())
```
For clarity, can you rename this function to `emitExecutableELFObject` ?

nbhuiyan commented on compiler/codegen/ELFGenerator.hpp:
```diff
@@ -0,0 +1,372 @@
+#include "env/TypedAllocator.hpp"
+#include "env/RawAllocator.hpp"
+#include "runtime/CodeCacheManager.hpp"
+//#include "codegen/StaticRelocation.hpp"
```
This would have been needed if the generator classes handled static relocations, which they don't as the task has been moved to CodeCacheManager. Definitely missed it during cleaning up.

nbhuiyan commented on compiler/codegen/ELFGenerator.cpp:
```diff
@@ -0,0 +1,562 @@
+}
+
+void
+TR::ELFGenerator::buildELFFile(::FILE *elfFile){
```
I thought of using the name `buildELFFile` mainly because it was putting all the pieces (building blocks) of the ELF file together. I think `emitELFFile` would be fine too.

nbhuiyan commented on compiler/codegen/ELFGenerator.cpp:
```diff
@@ -0,0 +1,562 @@
+TR::ELFGenerator::writeELFSymbolsToFile(::FILE *fp){
+    
+    ELFSymbol * elfSym = static_cast<ELFSymbol*>(_rawAllocator.allocate(sizeof(ELFSymbol)));
+    char ELFSymbolNames[_totalELFSymbolNamesLength];
```
Yes.

nbhuiyan commented on compiler/codegen/ELFGenerator.cpp:
```diff
@@ -0,0 +1,562 @@
+    }
+    
+    //values that are unchanged are being kept out of the while loop
+    elfSym->st_other = ELF64_ST_VISIBILITY(STV_DEFAULT);
```
Agreed.

nbhuiyan commented:
@0xdaryl I have addressed your comments. The code changes can be reviewed now. For now, I left the TR_PerfTool option enum and added TR_EnableExecutableELFFile option instead. Perhaps getting rid of this option can be in a separate PR.

nbhuiyan commented:
@0xdaryl This is no longer WIP. Requesting another review.

0xdaryl commented:
@genie-omr build all

0xdaryl commented on doc/compiler/runtime/ELFGenerator.md:
```diff
@@ -0,0 +1,183 @@
+generates relocatable ELF.
+
+### How to Enable ELF Generation
+The options `TR_EmitExecutableELFFile` and/or `TR_EmitRelocatableELFFile` will need
```
Can you provide the actual command-line option names rather than the enumerator list identifier?

0xdaryl commented on doc/compiler/runtime/ELFGenerator.md:
```diff
@@ -0,0 +1,183 @@
+option `objectFile=<object-file-name.o`. Relocatable ELF Files get generated in the
+working directory, and Executable ELF files are written to
+a file located in `/tmp/`, and named `perf-` followed by the process ID at the time
+of writing the ELF file. The relocatable ELF files are written to 
```
For the executable ELF files, perhaps mention that the choice of filename and location is largely historical.  (We should fix this up one day).

0xdaryl replied:
The last sentence trails off.  I think it can be removed anyway.

nbhuiyan commented:
@0xdaryl Thanks for the feedback on the documentation. I have made the changes you requested.

0xdaryl commented:
One more rebase please @nbhuiyan .

nbhuiyan commented:
@0xdaryl the merge conflicts are now fixed
